#2030 Add readOnly control in Common Properties

srikant-ch5 opened the pull request:
Fixes: #2030 
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

matthoward366 reviewed (commented):
Does this cover all control or only the carbon ones?  I think we need to cover even the controls that aren't carbon.

srikant-ch5 commented:
> Does this cover all control or only the carbon ones? I think we need to cover even the controls that aren't carbon.

@matthoward366 this covers only carbon ones as of now.
By controls that aren't carbon do you mean controls like `structureTable`, `readonlytable`, `expression`,`list`, etc. ?

matthoward366 commented:
>By controls that aren't carbon do you mean controls like structureTable, readonlytable, expression,list, etc. ?

Correct.  For tables it's probably just disabling the add/remove options then populating the readOnly down to the individual controls in the table.

srikant-ch5 commented:
Hi @matthoward366 I have added readonly in other controls as well please review.

srikant-ch5 commented:
Hi @matthoward366 In this PR the two controls for which I haven't included readonly are `structurelisteditor` and `structureTable` is it okay if we track this under a different issue or they be included in this same PR ?

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/controls/expression/expression.jsx:
```diff
@@ -442,7 +447,8 @@ ExpressionControl.propTypes = {
 	value: PropTypes.string, // pass in by redux
-	messageInfo: PropTypes.object // pass in by redux
+	messageInfo: PropTypes.object, // pass in by redux
+	readOnly: PropTypes.bool
```
Why do we have this and `this.props.control.readOnly`?

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/controls/list/list.jsx:
```diff
@@ -176,6 +179,7 @@ class ListControl extends AbstractTable {
 				updateRowSelections={this.updateRowSelections}
 				light={this.props.controller.getLight() && this.props.control.light}
 				emptyTablePlaceholder={this.props.control.additionalText}
+				readOnly={this.props.control.readOnly || this.props.readOnly}
```
This is really odd code.

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/controls/list/list.jsx:
```diff
@@ -37,6 +37,7 @@ class ListControl extends AbstractTable {
 		super(props);
 		this.addRow = this.addRow.bind(this);
 		this.reactIntl = props.controller.getReactIntl();
+		this.readOnly = this.props.control.readOnly;
```
Why have this?

srikant-ch5 replied:
This was used to disable to `Add Value` button in list. I have removed this and have used `this.props.readOnly` in child component.

<img width="301" alt="Screenshot 2024-07-16 at 4 30 03 PM" src="https://github.com/user-attachments/assets/62c0895b-5ebf-447d-8791-7b24b67dfe2b">

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/controls/numberfield/numberfield.jsx:
```diff
@@ -163,6 +163,7 @@ class NumberfieldControl extends React.Component {
 					allowEmpty
 					hideSteppers={this.props.tableControl || (this.props.control.controlType === ControlType.NUMBERFIELD)}
 					helperText={this.props.control.helperText}
+					readOnly={this.props.control.readOnly || this.props.readOnly}
```
Same as above.  

Seems like at the level when creating the control we should just always pass down the readOnly prop.  Then not use `this.props.control.readOnly` inside the controls.

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/controls/radioset/radioset.jsx:
```diff
@@ -115,12 +115,14 @@ class RadiosetControl extends React.Component {
-		const oldVal = this.props.controller.getPropertyValue(this.props.propertyId);
-		const newVal = this.convertTargetValue(evt);
-		this.props.controller.updatePropertyValue(this.props.propertyId, newVal);
+		if (!this.props.control.readOnly) {
```
Why do we need this if this should already be readOnly?

srikant-ch5 replied:
I have included this condition to disable switching between radio button. Please refer below recording in which we can switch between radio buttons even if readOnly is enabled.


https://github.com/user-attachments/assets/e2ba859c-1414-43ee-8af5-48443af3dcf1

srikant-ch5 commented:
Hi @matthoward366 Thank you for suggesting below approach.

> Seems like at the level when creating the control we should just always pass down the readOnly prop. Then not use this.props.control.readOnly inside the controls.

I have updated PR with all the controls using `this.props.readOnly` and have removed `this.props.control.readOnly`.

Regarding other feedback I have added my comments above respectively.

Would really appreciate your feedback on these changes, thanks.

srikant-ch5 commented:
I have fixed test cases for all controls to directly test using `readOnly` prop.

matthoward366 commented on docs/pages/04.03-ui-hints.md:
```diff
@@ -129,6 +129,7 @@ Parameter info attributes.
 * `display_chars` (*number*) This has been deprecated and is subject to removal. Limits the number of characters displayed for a text field in a column in a table.  The text will have an ellipsis appended at this limit. Defaults to 64 characters.
 * `separator` (*string*) Determines where to put a separator relative to the current control.  Valid values are ` after, before`.
 * `visible` (*boolean*) Determines whether to display control in a table cell. Used in complex types only.
+* `readOnly` (*boolean*) Determines whether the control should be immutable or can be edited.
```
This should be `read_only`

srikant-ch5 replied:
Fixed this and updated PR. Thanks.

srikant-ch5 commented:
Adding to my above note

> Note: I haven't included readonly controls in controls Structurelisteditor, StructureTable, Readonlytable as these three are based on abstract table, please let me know if these controls should also have readonly (except readonlytable I guess). If these controls need readOnly control then if readOnly is set to true do we only need to disable configure button ?

I have included readonly controls in Strucurelisteditor in a way where we click on Configure Button and then the table is readonly with button disabled as in below recording.

https://github.com/user-attachments/assets/e6a68568-ae1b-420e-886d-adb5e9d65960

This recording is of `Structurelisteditor` Could you please confirm if I should go ahead with same implementation for `StructureTable` and include these changes in this PR ?

matthoward366 reviewed (approved):
Changes look good.
